fs: compaction

tropuq opened the pull request:
This series is part of the ZPP FS project that is coordinated by Piotr Sarna <sarna@scylladb.com>.
The goal of this project is to create SeastarFS -- a fully asynchronous, sharded, user-space,
log-structured file system that is intended to become an alternative to XFS for Scylla.

SeastarFS is a log-structured file system. One of the characteristics of this file system is that data and metadata of every operation are appended to ends of the metadata log and data log. These logs are never modified. Because of that, some parts of logs may become out-dated after some time because of file modifications. SeastarFS must reclaim that out-dated space from these logs.

This patch introduces compaction -- a process of reclaiming clusters in SeastarFS. The process can be divided into two types: data compaction (reclaiming free clusters from data logs) and metadata compaction (reclaiming free clusters from metadata log). This patch focuses only on data compaction. Metadata log compaction won't be added in this series.

Data compaction consists of the following steps:
1. reading some number of clusters,
2. extracting active data from these clusters,
3. allocating new clusters for data,
4. writing data to new clusters,
5. freeing previous clusters to cluster allocator.

Example:
Let's say the cluster size is 1MB. Compaction reads 10 clusters, extracts 3MB of active data from these clusters, allocates 3 new clusters, and writes the data into them. This process reclaims 7 clusters.

Besides the implementation of the mentioned compaction process, this patch introduces some methods that decide when the compaction should be started and which clusters should be compacted.

wmitros committed:
fs: compaction: calculate metadata log and compacted metadata log sizes

Added new backned::shard members:
- _log_cluster_count - which holds the number of clusters on disk that
  span current metadata log;
- _compacted_log_size - which holds sum of sizes of the structs that
  represent the same data using possibly little space;

Implemented calculations that update these values.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

wmitros committed:
fs: compaction: add initial cluster metadata

Create a new class - data_cluster_contents_info - which holds info
about the data stored in a cluster.

Added new backend::shard members which enable finding corresponding
data_cluster_contents_info for every data cluster, as well as
discriminating the clusters that have been filled with data from ones
that are in the middle of this process.

Implemented calculations that update the data stored in those metadata
log fields.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

tropuq committed:
fs: compaction: added rwlocks for data_clusters

Read locks will be used to indicate that someone is reading
from the given cluster. Write locks will be used in compaction
before releasing compacted clusters. We need to wait for all
readers to end reading from the given cluster before releasing
it to the cluster_allocator.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
fs: compaction: added implementation of data clusters compaction

Introduced a method that executes a compaction. The method accepts a
list of read-only clusters and performs compaction on these clusters.

The compaction process
1. Read compacted clusters and store active (not out-dated) data in
   data vectors
2. Allocate new clusters needed to store compacted data
3. Write data vectors into new clusters and update positions of the
   data in memory
4. Return free clusters to cluster allocator.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
fs: cluster_allocator: add method returning future waiting for clusters

* Added method alloc_wait() that will be used in compaction to allocate
  all needed clusters with one call.
* Code was refactored to provide a strong exception guarantee.
* Tests don't assume the order of returned clusters.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
fs: backend: add read-only mode

Added private function for switching filesystem to read-only mode.
It will be used in emergency situations like failing flush_log().

Signed-off-by: Michał Niciejewski <quport@gmail.com>

wmitros committed:
fs: compaction: data compaction execution

The decision on which clusters to compact and when should it be done is
based on two variables: "compactness" and "max_data_compaction_memory",
which are set as parameters of the filesystem. The compaction is
performed when enough clusters contain little up-to-date data; enough
clusters meaning so much that including another one would require using
more memory than max_data_compaction_memory, and little data meaning
less than compactness*size of cluster.
This patch adds those variables and the decision making that is based
on them.

Signed-off-by: Wojciech Mitros <wmitros@protonmail.com>

psarna commented:
Looks good, please rebase on top of our branch and feel free to merge. (if there are leftover fixup commits, let's squash them)

tropuq committed:
fs: cluster_allocator: added logs

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
fs: compaction: add timeout for waiting for clusters

Timeout is helpful in cases of using the system with low remaining
space. In this case, compaction won't wait forever for new clusters but
instead return an error after some time.

Signed-off-by: Michał Niciejewski <quport@gmail.com>

tropuq committed:
fs: compaction: added FIXME in shutdown method

Signed-off-by: Michał Niciejewski <quport@gmail.com>
